File: src/jastudio/note/vocabulary/related_vocab/Antonyms.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from jastudio.ankiutils import app
from jastudio.ankiutils.app import col
from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]

if TYPE_CHECKING:
    from note.notefields.json_object_field import MutableSerializedObjectField
    from note.vocabulary.related_vocab.related_vocab_data import RelatedVocabData
    from note.vocabulary.vocabnote import VocabNote
    from sysutils.weak_ref import WeakRef
    from typed_linq_collections.collections.q_set import QSet

class Antonyms(Slots):
    def __init__(self, vocab: WeakRef[VocabNote], data: MutableSerializedObjectField[RelatedVocabData]) -> None:
        self._vocab: WeakRef[VocabNote] = vocab
        self._data: MutableSerializedObjectField[RelatedVocabData] = data

    def strings(self) -> QSet[str]: return self._data.get().antonyms
    def notes(self) -> list[VocabNote]:
        return col().vocab.with_any_form_in_prefer_disambiguation_name_or_exact_match(list(self.strings()))

    def add(self, antonym: str) -> None:
        self.strings().add(antonym)

        for similar in app.col().vocab.with_question(antonym):
            if self._vocab().get_question() not in similar.related_notes.antonyms.strings():
                similar.related_notes.antonyms.add(self._vocab().get_question())

        self._data.save()

    def remove(self, to_remove: str) -> None:
        self.strings().remove(to_remove)

        for similar in app.col().vocab.with_question(to_remove):
            if self._vocab().get_question() in similar.related_notes.antonyms.strings():
                similar.related_notes.antonyms.remove(self._vocab().get_question())

        self._data.save()

    @override
    def __repr__(self) -> str: return self._data.__repr__()

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/RelatedVocab/Antonyms.cs
using System.Collections.Generic;
using System.Linq;

namespace JAStudio.Core.Note.Vocabulary.RelatedVocab;

public class Antonyms
{
   readonly VocabNote _vocab;
   readonly RelatedVocabData _data;
   readonly NoteGuard _guard;

   public Antonyms(VocabNote vocab, RelatedVocabData data, NoteGuard guard)
   {
      _vocab = vocab;
      _data = data;
      _guard = guard;
   }

   public HashSet<string> Strings() => _data.Antonyms;

   public List<VocabNote> Notes() => _vocab.Services.Collection.Vocab.WithAnyFormInPreferDisambiguationNameOrExactMatch(Strings().ToList());

   public void Add(string antonym) => _guard.Update(() =>
   {
      Strings().Add(antonym);

      foreach(var similar in _vocab.Services.Collection.Vocab.WithQuestion(antonym))
      {
         if(!similar.RelatedNotes.Antonyms.Strings().Contains(_vocab.GetQuestion()))
         {
            similar.RelatedNotes.Antonyms.Add(_vocab.GetQuestion());
         }
      }
   });

   public void Remove(string toRemove) => _guard.Update(() =>
   {
      Strings().Remove(toRemove);

      foreach(var similar in _vocab.Services.Collection.Vocab.WithQuestion(toRemove))
      {
         if(similar.RelatedNotes.Antonyms.Strings().Contains(_vocab.GetQuestion()))
         {
            similar.RelatedNotes.Antonyms.Remove(_vocab.GetQuestion());
         }
      }
   });

   public override string ToString() => string.Join(", ", Strings());
}
